### PR TITLE
Refactor more methods to the SearchService

### PR DESCRIPTION
**Pull request by demiankatz**
This fills in more methods of the search service. Note that some of the work here is redundant with #3732 because I had to copy some test refactoring from there to get things passing here. If that PR is reviewed first, it will reduce the diffs here. Also note that I haven't updated the expiration command yet; that will be easier after #3714 is merged.

TODO
- [x] Update changelog when merging (VuFind\Search\SearchNormalizer and VuFindConsole\Command\ScheduledSearch\NotifyCommand constructors)

### CODE REVIEW

File: module/VuFind/src/VuFind/Crypt/SecretCalculator.php [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+/**
+ * Secret calculator
+ *
+ * @category VuFind
+ * @package  Crypt
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+class SecretCalculator
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Crypt/SecretCalculator.php:43]: I don't love this class, but the search unsubscribe calculation had to go into a shared location, and it seemed better to create a new class than to overload the existing HMAC class, which seemed like inappropriate scope creep.

File: module/VuFindConsole/src/VuFindConsole/Command/ScheduledSearch/NotifyCommand.php [hunk 2]
```diff
@@ -294,35 +221,24 @@ protected function validateSchedule($todayTime, $lastTime, $s)
...
-        if ($user === false || $s->user_id != $user->getId()) {
-            if (!$user = $this->userTable->getById($s->user_id)) {
-                $user = false;  // make sure static variable is cleared
-                $this->warn(
-                    'Search ' . $s->id . ': user ' . $s->user_id
-                    . ' does not exist '
-                );
-                return false;
-            }
+        if (!$user = $s->getUser()) {
```
Comment (demiankatz) [inline module/VuFindConsole/src/VuFindConsole/Command/ScheduledSearch/NotifyCommand.php:233]: Due to changes in the way the search and user entities relate to one another, it seemed cleanest to eliminate the user caching logic in order to simplify the code and reduce dependencies. In some situations, this may put slightly higher load on the database, but looking up a user by primary key is not an expensive operation, so I'm not too concerned about it.

File: module/VuFindConsole/tests/unit-tests/src/VuFindTest/Command/ScheduledSearch/NotifyCommandTest.php [hunk 3]
```diff
@@ -139,55 +143,26 @@ public function testNotificationWithRecentExecution()
         $this->assertEquals(0, $commandTester->getStatusCode());
     }
 
-    /**
-     * Test behavior when notifications are waiting to be sent but user does not
-     * exist.
-     *
-     * @return void
-     */
-    public function testNotificationsWithMissingUser()
```
Comment (demiankatz) [inline module/VuFindConsole/tests/unit-tests/src/VuFindTest/Command/ScheduledSearch/NotifyCommandTest.php:148]: Due to changes in user handling in the command, this test no longer applies, so I simply deleted it.

File: module/VuFind/src/VuFind/Crypt/SecretCalculator.php (hunk 1 above)
Comment (aleksip) [reply module/VuFind/src/VuFind/Crypt/SecretCalculator.php:43]: Well at least the name of the class made me smile. 😄

File: module/VuFind/src/VuFind/Db/Service/SearchServiceInterface.php [hunk 4]
```diff
@@ -40,10 +43,104 @@
...
+     */
+    public function getScheduledSearches(): array;
+
+    /**
+     * Retrieve all searches matching the specified checksum and belonging to the user specified by session or user
+     * entity/ID.
+     *
+     * @param int                          $checksum  Checksum to match
+     * @param string                       $sessionId Current session ID
+     * @param null|UserEntityInterface|int $userOrId  Entity or ID representing current user.
```
Comment (aleksip) [inline module/VuFind/src/VuFind/Db/Service/SearchServiceInterface.php]: Nitpick: type order and mention about being optional.

File: module/VuFind/src/VuFind/Search/SearchNormalizer.php [hunk 5]
```diff
@@ -83,4 +84,98 @@ public function normalizeMinifiedSearch(Minified $minified): NormalizedSearch
...
+     * @param int|null                    $userId    Current user ID
+     *
+     * @return SearchEntityInterface
+     * @throws Exception
+     */
+    public function saveNormalizedSearch(
+        $results,
+        $sessionId,
+        $userId
+    ) {
```
Comment (aleksip) [inline module/VuFind/src/VuFind/Search/SearchNormalizer.php]: Parameter and return types?

File: module/VuFind/tests/unit-tests/src/VuFindTest/Session/DatabaseTest.php [hunk 6]
```diff
@@ -109,7 +109,7 @@ public function testWrite(): void
      *
      * @return void
      */
-    public function testDestroy(): void
+    public function ztestDestroy(): void
```
Comment (aleksip) [inline module/VuFind/tests/unit-tests/src/VuFindTest/Session/DatabaseTest.php]: ?

**Review by aleksip — APPROVED** [review]

File: module/VuFind/tests/unit-tests/src/VuFindTest/Session/DatabaseTest.php (hunk 6 above)
Comment (demiankatz) [reply module/VuFind/tests/unit-tests/src/VuFindTest/Session/DatabaseTest.php]: This is how I disable tests, but no idea why I disabled that and forgot to re-enable it. Definitely not intentional; reverted now (and the test is passing :-) ).
